File: java/src/main/java/tech/hearth/crypto/SodiumBackend.java
package tech.hearth.crypto;

import java.lang.foreign.Arena;
import java.lang.foreign.FunctionDescriptor;
import java.lang.foreign.Linker;
import java.lang.foreign.MemorySegment;
import java.lang.foreign.SymbolLookup;
import java.lang.foreign.ValueLayout;
import java.lang.invoke.MethodHandle;
import java.util.List;
import java.util.Optional;

/**
 * libsodium binding via the Java Foreign Function &amp; Memory API (Panama). No
 * JNI and no extra native artifacts — it {@code dlopen}s the platform libsodium
 * and drives it directly. Constructing the backend loads the library; if that
 * fails the constructor throws and callers fall back to {@link JvmBackend}.
 */
public final class SodiumBackend implements CryptoBackend {

    private static final ValueLayout.OfInt INT = ValueLayout.JAVA_INT;
    private static final ValueLayout.OfLong LONG = ValueLayout.JAVA_LONG;
    private static final java.lang.foreign.AddressLayout PTR = ValueLayout.ADDRESS;

    private static final int SIGN_PUBLICKEY_BYTES = 32;
    private static final int SIGN_SECRETKEY_BYTES = 64;
    private static final int SIGN_BYTES = 64;
    private static final int SHA512_BYTES = 64;
    private static final int SHA256_BYTES = 32;
    private static final int POINT_BYTES = 32;
    private static final int SCALAR_BYTES = 32;

    private final Linker linker = Linker.nativeLinker();
    private final SymbolLookup lookup;

    private final MethodHandle hSeedKeypair;
    private final MethodHandle hSignDetached;
    private final MethodHandle hVerifyDetached;
    private final MethodHandle hSha512;
    private final MethodHandle hSha256;
    private final MethodHandle hPointAdd;
    private final MethodHandle hPointSub;
    private final MethodHandle hScalarmultNoclamp;
    private final MethodHandle hScalarmultBaseNC;
    private final MethodHandle hScalarMul;
    private final MethodHandle hScalarAdd;
    private final MethodHandle hScalarReduce;
    private final MethodHandle hHmac512Statebytes;
    private final MethodHandle hHmac512Init;
    private final MethodHandle hHmac512Update;
    private final MethodHandle hHmac512Final;
    private final MethodHandle hHmac256Statebytes;
    private final MethodHandle hHmac256Init;
    private final MethodHandle hHmac256Update;
    private final MethodHandle hHmac256Final;

    public SodiumBackend() {
        this.lookup = loadLibrary();
        this.hSeedKeypair = handle("crypto_sign_seed_keypair", FunctionDescriptor.of(INT, PTR, PTR, PTR));
        this.hSignDetached = handle("crypto_sign_detached", FunctionDescriptor.of(INT, PTR, PTR, PTR, LONG, PTR));
        this.hVerifyDetached = handle("crypto_sign_verify_detached", FunctionDescriptor.of(INT, PTR, PTR, LONG, PTR));
        this.hSha512 = handle("crypto_hash_sha512", FunctionDescriptor.of(INT, PTR, PTR, LONG));
        this.hSha256 = handle("crypto_hash_sha256", FunctionDescriptor.of(INT, PTR, PTR, LONG));
        this.hPointAdd = handle("crypto_core_ed25519_add", FunctionDescriptor.of(INT, PTR, PTR, PTR));
        this.hPointSub = handle("crypto_core_ed25519_sub", FunctionDescriptor.of(INT, PTR, PTR, PTR));
        this.hScalarmultNoclamp = handle("crypto_scalarmult_ed25519_noclamp", FunctionDescriptor.of(INT, PTR, PTR, PTR));
        this.hScalarmultBaseNC = handle("crypto_scalarmult_ed25519_base_noclamp", FunctionDescriptor.of(INT, PTR, PTR));
        this.hScalarMul = handle("crypto_core_ed25519_scalar_mul", FunctionDescriptor.ofVoid(PTR, PTR, PTR));
        this.hScalarAdd = handle("crypto_core_ed25519_scalar_add", FunctionDescriptor.ofVoid(PTR, PTR, PTR));
        this.hScalarReduce = handle("crypto_core_ed25519_scalar_reduce", FunctionDescriptor.ofVoid(PTR, PTR));
        this.hHmac512Statebytes = handle("crypto_auth_hmacsha512_statebytes", FunctionDescriptor.of(LONG));
        this.hHmac512Init = handle("crypto_auth_hmacsha512_init", FunctionDescriptor.of(INT, PTR, PTR, LONG));
        this.hHmac512Update = handle("crypto_auth_hmacsha512_update", FunctionDescriptor.of(INT, PTR, PTR, LONG));
        this.hHmac512Final = handle("crypto_auth_hmacsha512_final", FunctionDescriptor.of(INT, PTR, PTR));
        this.hHmac256Statebytes = handle("crypto_auth_hmacsha256_statebytes", FunctionDescriptor.of(LONG));
        this.hHmac256Init = handle("crypto_auth_hmacsha256_init", FunctionDescriptor.of(INT, PTR, PTR, LONG));
        this.hHmac256Update = handle("crypto_auth_hmacsha256_update", FunctionDescriptor.of(INT, PTR, PTR, LONG));
        this.hHmac256Final = handle("crypto_auth_hmacsha256_final", FunctionDescriptor.of(INT, PTR, PTR));

        MethodHandle init = handle("sodium_init", FunctionDescriptor.of(INT));
        if (callInt(init) < 0) {
            throw new IllegalStateException("sodium_init() failed");
        }
    }

    private SymbolLookup loadLibrary() {
        List<String> candidates = List.of(
                System.getenv().getOrDefault("HEARTH_SODIUM_LIB", ""),
                "libsodium.dylib",
                "libsodium.so.23",
                "libsodium.so",
                "/opt/homebrew/lib/libsodium.dylib",
                "/usr/local/lib/libsodium.dylib",
                "/usr/lib/x86_64-linux-gnu/libsodium.so.23");
        for (String name : candidates) {
            if (name.isEmpty()) {
                continue;
            }
            try {
                return SymbolLookup.libraryLookup(name, Arena.global());
            } catch (RuntimeException ignored) {
                // try next candidate
            }
        }
        throw new UnsatisfiedLinkError(
                "Could not load libsodium. Install it (macOS: `brew install libsodium`, "
                        + "Debian/Ubuntu: `apt install libsodium23`) or set HEARTH_SODIUM_LIB.");
    }

    private MethodHandle handle(String name, FunctionDescriptor desc) {
        MemorySegment sym = lookup.find(name)
                .orElseThrow(() -> new UnsatisfiedLinkError("libsodium symbol not found: " + name));
        return linker.downcallHandle(sym, desc);
    }

    /** Force native loading + a trivial call so callers can probe availability. */
    public void selfTest() {
        sha512(new byte[0]);
    }

    @Override
    public String name() {
        return "libsodium";
    }

    @Override
    public byte[] sha512(byte[] in) {
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment out = arena.allocate(SHA512_BYTES);
            callInt(hSha512, out, seg(arena, in), Long.valueOf(in.length));
            return read(out, SHA512_BYTES);
        }
    }

    @Override
    public byte[] sha256(byte[] in) {
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment out = arena.allocate(SHA256_BYTES);
            callInt(hSha256, out, seg(arena, in), Long.valueOf(in.length));
            return read(out, SHA256_BYTES);
        }
    }

    @Override
    public RawKeypair signSeedKeypair(byte[] seed) {
        if (seed.length != 32) {
            throw new IllegalArgumentException("seed must be 32 bytes");
        }
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment pk = arena.allocate(SIGN_PUBLICKEY_BYTES);
            MemorySegment sk = arena.allocate(SIGN_SECRETKEY_BYTES);
            if (callInt(hSeedKeypair, pk, sk, seg(arena, seed)) != 0) {
                throw new RuntimeException("crypto_sign_seed_keypair failed");
            }
            return new RawKeypair(read(pk, SIGN_PUBLICKEY_BYTES), read(sk, SIGN_SECRETKEY_BYTES));
        }
    }

    @Override
    public byte[] signDetached(byte[] msg, byte[] secretKey) {
        if (secretKey.length != SIGN_SECRETKEY_BYTES) {
            throw new IllegalArgumentException("secret key must be 64 bytes");
        }
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment sig = arena.allocate(SIGN_BYTES);
            int rc = callInt(hSignDetached, sig, MemorySegment.NULL, seg(arena, msg),
                    Long.valueOf(msg.length), seg(arena, secretKey));
            if (rc != 0) {
                throw new RuntimeException("crypto_sign_detached failed");
            }
            return read(sig, SIGN_BYTES);
        }
    }

    @Override
    public boolean verifyDetached(byte[] sig, byte[] msg, byte[] publicKey) {
        if (sig.length != SIGN_BYTES || publicKey.length != SIGN_PUBLICKEY_BYTES) {
            return false;
        }
        try (Arena arena = Arena.ofConfined()) {
            return callInt(hVerifyDetached, seg(arena, sig), seg(arena, msg),
                    Long.valueOf(msg.length), seg(arena, publicKey)) == 0;
        }
    }

    @Override
    public Optional<byte[]> pointAdd(byte[] p, byte[] q) {
        return twoPointOp(hPointAdd, p, q);
    }

    @Override
    public Optional<byte[]> pointSub(byte[] p, byte[] q) {
        return twoPointOp(hPointSub, p, q);
    }

    private Optional<byte[]> twoPointOp(MethodHandle op, byte[] p, byte[] q) {
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment r = arena.allocate(POINT_BYTES);
            int rc = callInt(op, r, seg(arena, p), seg(arena, q));
            return rc == 0 ? Optional.of(read(r, POINT_BYTES)) : Optional.empty();
        }
    }

    @Override
    public Optional<byte[]> scalarmultNoclamp(byte[] n, byte[] p) {
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment out = arena.allocate(POINT_BYTES);
            int rc = callInt(hScalarmultNoclamp, out, seg(arena, n), seg(arena, p));
            return rc == 0 ? Optional.of(read(out, POINT_BYTES)) : Optional.empty();
        }
    }

    @Override
    public byte[] scalarmultBaseNoclamp(byte[] n) {
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment out = arena.allocate(POINT_BYTES);
            if (callInt(hScalarmultBaseNC, out, seg(arena, n)) != 0) {
                throw new RuntimeException("crypto_scalarmult_ed25519_base_noclamp failed");
            }
            return read(out, POINT_BYTES);
        }
    }

    @Override
    public byte[] scalarMul(byte[] x, byte[] y) {
        return scalarBinOp(hScalarMul, x, y);
    }

    @Override
    public byte[] scalarAdd(byte[] x, byte[] y) {
        return scalarBinOp(hScalarAdd, x, y);
    }

    private byte[] scalarBinOp(MethodHandle op, byte[] x, byte[] y) {
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment z = arena.allocate(SCALAR_BYTES);
            callVoid(op, z, seg(arena, x), seg(arena, y));
            return read(z, SCALAR_BYTES);
        }
    }

    @Override
    public byte[] scalarReduce(byte[] wide) {
        if (wide.length != 64) {
            throw new IllegalArgumentException("input must be 64 bytes");
        }
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment r = arena.allocate(SCALAR_BYTES);
            callVoid(hScalarReduce, r, seg(arena, wide));
            return read(r, SCALAR_BYTES);
        }
    }

    @Override
    public byte[] hmacSha512(byte[] key, byte[] msg) {
        return hmac(hHmac512Statebytes, hHmac512Init, hHmac512Update, hHmac512Final, SHA512_BYTES, key, msg);
    }

    @Override
    public byte[] hmacSha256(byte[] key, byte[] msg) {
        return hmac(hHmac256Statebytes, hHmac256Init, hHmac256Update, hHmac256Final, SHA256_BYTES, key, msg);
    }

    private byte[] hmac(MethodHandle statebytes, MethodHandle init, MethodHandle update, MethodHandle fin,
                        int outLen, byte[] key, byte[] msg) {
        try (Arena arena = Arena.ofConfined()) {
            long stateLen = callLong(statebytes);
            MemorySegment state = arena.allocate(stateLen, 16);
            if (callInt(init, state, seg(arena, key), Long.valueOf(key.length)) != 0) {
                throw new RuntimeException("hmac init failed");
            }
            if (callInt(update, state, seg(arena, msg), Long.valueOf(msg.length)) != 0) {
                throw new RuntimeException("hmac update failed");
            }
            MemorySegment out = arena.allocate(outLen);
            if (callInt(fin, state, out) != 0) {
                throw new RuntimeException("hmac final failed");
            }
            return read(out, outLen);
        }
    }

    // --- FFI helpers ---------------------------------------------------------

    private static MemorySegment seg(Arena arena, byte[] bytes) {
        MemorySegment s = arena.allocate(Math.max(bytes.length, 1));
        MemorySegment.copy(bytes, 0, s, ValueLayout.JAVA_BYTE, 0L, bytes.length);
        return s;
    }

    private static byte[] read(MemorySegment s, int len) {
        byte[] a = new byte[len];
        MemorySegment.copy(s, ValueLayout.JAVA_BYTE, 0L, a, 0, len);
        return a;
    }

    private static int callInt(MethodHandle h, Object... args) {
        try {
            return (Integer) h.invokeWithArguments(args);
        } catch (Throwable t) {
            throw new RuntimeException(t);
        }
    }

    private static long callLong(MethodHandle h, Object... args) {
        try {
            return (Long) h.invokeWithArguments(args);
        } catch (Throwable t) {
            throw new RuntimeException(t);
        }
    }

    private static void callVoid(MethodHandle h, Object... args) {
        try {
            h.invokeWithArguments(args);
        } catch (Throwable t) {
            throw new RuntimeException(t);
        }
    }
}

File: scala/src/main/scala/hearth/crypto/SodiumBackend.scala
package hearth.crypto

import java.lang.foreign.*
import java.lang.invoke.MethodHandle

/** Thin binding to libsodium via the Java 22+ Foreign Function & Memory API
  * (Project Panama). No JNI, no extra native artifacts to ship — we just
  * `dlopen` the platform libsodium and drive it directly.
  *
  * Every method here maps 1:1 onto a libsodium C entry point and works purely
  * on `Array[Byte]`; the segment/arena bookkeeping stays inside this file.
  */
object SodiumBackend extends CryptoBackend:

  def name: String = "libsodium"

  /** Force native loading + a trivial call so callers can probe availability. */
  def selfTest(): Unit = { sha512(Array.emptyByteArray); () }

  // --- Constants (from sodium headers, stable across releases) -------------
  private val SignPublicKeyBytes: Int = 32 // crypto_sign_PUBLICKEYBYTES
  private val SignSecretKeyBytes: Int = 64 // crypto_sign_SECRETKEYBYTES (seed||pk)
  private val SignBytes: Int          = 64 // crypto_sign_BYTES
  private val Sha512Bytes: Int        = 64
  private val Sha256Bytes: Int        = 32
  private val PointBytes: Int         = 32 // crypto_core_ed25519_BYTES
  private val ScalarBytes: Int        = 32 // crypto_core_ed25519_SCALARBYTES
  private val ScalarNonReducedBytes: Int = 64 // crypto_core_ed25519_NONREDUCEDSCALARBYTES
  private val SignSeedBytes: Int      = 32 // crypto_sign_SEEDBYTES

  private val linker: Linker = Linker.nativeLinker()
  private val lookup: SymbolLookup =
    val candidates = List(
      Option(System.getenv("HEARTH_SODIUM_LIB")),
      Some("libsodium.dylib"),
      Some("libsodium.so.23"),
      Some("libsodium.so"),
      Some("/opt/homebrew/lib/libsodium.dylib"),
      Some("/usr/local/lib/libsodium.dylib"),
      Some("/usr/lib/x86_64-linux-gnu/libsodium.so.23")
    ).flatten
    candidates.iterator
      .flatMap(name => scala.util.Try(SymbolLookup.libraryLookup(name, Arena.global())).toOption)
      .nextOption()
      .getOrElse(
        throw new UnsatisfiedLinkError(
          "Could not load libsodium. Install it (macOS: `brew install libsodium`, " +
            "Debian/Ubuntu: `apt install libsodium23`) or set HEARTH_SODIUM_LIB to the .dylib/.so path."
        )
      )

  private def handle(name: String, desc: FunctionDescriptor): MethodHandle =
    val sym = lookup
      .find(name)
      .orElseThrow(() => new UnsatisfiedLinkError(s"libsodium symbol not found: $name"))
    linker.downcallHandle(sym, desc)

  private val PTR  = ValueLayout.ADDRESS
  private val INT  = ValueLayout.JAVA_INT
  private val LONG = ValueLayout.JAVA_LONG // size_t / unsigned long long on LP64

  private val hInit               = handle("sodium_init", FunctionDescriptor.of(INT))
  private val hSeedKeypair        = handle("crypto_sign_seed_keypair", FunctionDescriptor.of(INT, PTR, PTR, PTR))
  private val hSignDetached       = handle("crypto_sign_detached", FunctionDescriptor.of(INT, PTR, PTR, PTR, LONG, PTR))
  private val hVerifyDetached     = handle("crypto_sign_verify_detached", FunctionDescriptor.of(INT, PTR, PTR, LONG, PTR))
  private val hSha512             = handle("crypto_hash_sha512", FunctionDescriptor.of(INT, PTR, PTR, LONG))
  private val hSha256             = handle("crypto_hash_sha256", FunctionDescriptor.of(INT, PTR, PTR, LONG))
  private val hPointAdd           = handle("crypto_core_ed25519_add", FunctionDescriptor.of(INT, PTR, PTR, PTR))
  private val hPointSub           = handle("crypto_core_ed25519_sub", FunctionDescriptor.of(INT, PTR, PTR, PTR))
  private val hScalarmultNoclamp  = handle("crypto_scalarmult_ed25519_noclamp", FunctionDescriptor.of(INT, PTR, PTR, PTR))
  private val hScalarmultBaseNC   = handle("crypto_scalarmult_ed25519_base_noclamp", FunctionDescriptor.of(INT, PTR, PTR))
  private val hScalarMul          = handle("crypto_core_ed25519_scalar_mul", FunctionDescriptor.ofVoid(PTR, PTR, PTR))
  private val hScalarAdd          = handle("crypto_core_ed25519_scalar_add", FunctionDescriptor.ofVoid(PTR, PTR, PTR))
  private val hScalarReduce       = handle("crypto_core_ed25519_scalar_reduce", FunctionDescriptor.ofVoid(PTR, PTR))
  private val hHmac512Statebytes  = handle("crypto_auth_hmacsha512_statebytes", FunctionDescriptor.of(LONG))
  private val hHmac512Init        = handle("crypto_auth_hmacsha512_init", FunctionDescriptor.of(INT, PTR, PTR, LONG))
  private val hHmac512Update      = handle("crypto_auth_hmacsha512_update", FunctionDescriptor.of(INT, PTR, PTR, LONG))
  private val hHmac512Final       = handle("crypto_auth_hmacsha512_final", FunctionDescriptor.of(INT, PTR, PTR))
  private val hHmac256Statebytes  = handle("crypto_auth_hmacsha256_statebytes", FunctionDescriptor.of(LONG))
  private val hHmac256Init        = handle("crypto_auth_hmacsha256_init", FunctionDescriptor.of(INT, PTR, PTR, LONG))
  private val hHmac256Update      = handle("crypto_auth_hmacsha256_update", FunctionDescriptor.of(INT, PTR, PTR, LONG))
  private val hHmac256Final       = handle("crypto_auth_hmacsha256_final", FunctionDescriptor.of(INT, PTR, PTR))

  // sodium_init() is idempotent and thread-safe; must be called before use.
  private val initResult: Int = hInit.invokeWithArguments().asInstanceOf[Integer].intValue
  if initResult < 0 then throw new IllegalStateException("sodium_init() failed")

  // --- Small helpers -------------------------------------------------------

  private inline def withArena[A](f: Arena => A): A =
    val arena = Arena.ofConfined()
    try f(arena)
    finally arena.close()

  private def seg(arena: Arena, bytes: Array[Byte]): MemorySegment =
    val s = arena.allocate(bytes.length.toLong.max(1L))
    MemorySegment.copy(bytes, 0, s, ValueLayout.JAVA_BYTE, 0L, bytes.length)
    s

  private def out(arena: Arena, len: Int): MemorySegment = arena.allocate(len.toLong)

  private def read(s: MemorySegment, len: Int): Array[Byte] =
    val a = new Array[Byte](len)
    MemorySegment.copy(s, ValueLayout.JAVA_BYTE, 0L, a, 0, len)
    a

  // --- Public API ----------------------------------------------------------

  /** Deterministically expand a 32-byte seed into an Ed25519 keypair.
    * Returns (publicKey[32], secretKey[64] = seed||publicKey).
    */
  def signSeedKeypair(seed: Array[Byte]): (Array[Byte], Array[Byte]) =
    require(seed.length == SignSeedBytes, "seed must be 32 bytes")
    withArena { a =>
      val pk = out(a, SignPublicKeyBytes)
      val sk = out(a, SignSecretKeyBytes)
      val rc = hSeedKeypair.invokeWithArguments(pk, sk, seg(a, seed)).asInstanceOf[Integer].intValue
      if rc != 0 then throw new RuntimeException("crypto_sign_seed_keypair failed")
      (read(pk, SignPublicKeyBytes), read(sk, SignSecretKeyBytes))
    }

  /** Detached Ed25519 (EdDSA / RFC 8032) signature over `msg`. */
  def signDetached(msg: Array[Byte], secretKey: Array[Byte]): Array[Byte] =
    require(secretKey.length == SignSecretKeyBytes, "secret key must be 64 bytes")
    withArena { a =>
      val sig = out(a, SignBytes)
      val rc = hSignDetached
        .invokeWithArguments(sig, MemorySegment.NULL, seg(a, msg), java.lang.Long.valueOf(msg.length.toLong), seg(a, secretKey))
        .asInstanceOf[Integer].intValue
      if rc != 0 then throw new RuntimeException("crypto_sign_detached failed")
      read(sig, SignBytes)
    }

  /** Verify a detached Ed25519 signature. */
  def verifyDetached(sig: Array[Byte], msg: Array[Byte], publicKey: Array[Byte]): Boolean =
    require(sig.length == SignBytes, "signature must be 64 bytes")
    require(publicKey.length == SignPublicKeyBytes, "public key must be 32 bytes")
    withArena { a =>
      hVerifyDetached
        .invokeWithArguments(seg(a, sig), seg(a, msg), java.lang.Long.valueOf(msg.length.toLong), seg(a, publicKey))
        .asInstanceOf[Integer].intValue == 0
    }

  def sha512(in: Array[Byte]): Array[Byte] = withArena { a =>
    val o = out(a, Sha512Bytes)
    hSha512.invokeWithArguments(o, seg(a, in), java.lang.Long.valueOf(in.length.toLong))
    read(o, Sha512Bytes)
  }

  def sha256(in: Array[Byte]): Array[Byte] = withArena { a =>
    val o = out(a, Sha256Bytes)
    hSha256.invokeWithArguments(o, seg(a, in), java.lang.Long.valueOf(in.length.toLong))
    read(o, Sha256Bytes)
  }

  /** Ed25519 point addition on compressed points. Only checks the operands are
    * on the curve (not subgroup / small-order), which is exactly what RFC 9381
    * try-and-increment cofactor clearing needs. Returns None if off-curve.
    */
  def pointAdd(p: Array[Byte], q: Array[Byte]): Option[Array[Byte]] = withArena { a =>
    val r = out(a, PointBytes)
    val rc = hPointAdd.invokeWithArguments(r, seg(a, p), seg(a, q)).asInstanceOf[Integer].intValue
    if rc == 0 then Some(read(r, PointBytes)) else None
  }

  /** Ed25519 point subtraction p - q on compressed points (on-curve check only). */
  def pointSub(p: Array[Byte], q: Array[Byte]): Option[Array[Byte]] = withArena { a =>
    val r = out(a, PointBytes)
    val rc = hPointSub.invokeWithArguments(r, seg(a, p), seg(a, q)).asInstanceOf[Integer].intValue
    if rc == 0 then Some(read(r, PointBytes)) else None
  }

  /** q = n * p on Ed25519, scalar used verbatim (no clamping). Requires p on the
    * prime-order subgroup (libsodium enforces this). Returns None if rejected.
    */
  def scalarmultNoclamp(n: Array[Byte], p: Array[Byte]): Option[Array[Byte]] = withArena { a =>
    val q = out(a, PointBytes)
    val rc = hScalarmultNoclamp.invokeWithArguments(q, seg(a, n), seg(a, p)).asInstanceOf[Integer].intValue
    if rc == 0 then Some(read(q, PointBytes)) else None
  }

  /** q = n * B on Ed25519, scalar used verbatim (no clamping). */
  def scalarmultBaseNoclamp(n: Array[Byte]): Array[Byte] = withArena { a =>
    val q = out(a, PointBytes)
    val rc = hScalarmultBaseNC.invokeWithArguments(q, seg(a, n)).asInstanceOf[Integer].intValue
    if rc != 0 then throw new RuntimeException("crypto_scalarmult_ed25519_base_noclamp failed")
    read(q, PointBytes)
  }

  /** z = x * y mod L (group order). Inputs are 32-byte little-endian scalars. */
  def scalarMul(x: Array[Byte], y: Array[Byte]): Array[Byte] = withArena { a =>
    val z = out(a, ScalarBytes)
    hScalarMul.invokeWithArguments(z, seg(a, x), seg(a, y))
    read(z, ScalarBytes)
  }

  /** z = x + y mod L. */
  def scalarAdd(x: Array[Byte], y: Array[Byte]): Array[Byte] = withArena { a =>
    val z = out(a, ScalarBytes)
    hScalarAdd.invokeWithArguments(z, seg(a, x), seg(a, y))
    read(z, ScalarBytes)
  }

  /** Reduce a 64-byte little-endian value mod L to a 32-byte scalar. */
  def scalarReduce(wide: Array[Byte]): Array[Byte] =
    require(wide.length == ScalarNonReducedBytes, "input must be 64 bytes")
    withArena { a =>
      val r = out(a, ScalarBytes)
      hScalarReduce.invokeWithArguments(r, seg(a, wide))
      read(r, ScalarBytes)
    }

  /** HMAC-SHA-512 with an arbitrary-length key (the streaming API accepts any
    * key length, unlike the one-shot `crypto_auth_hmacsha512`).
    */
  def hmacSha512(key: Array[Byte], msg: Array[Byte]): Array[Byte] =
    hmac(hHmac512Statebytes, hHmac512Init, hHmac512Update, hHmac512Final, Sha512Bytes, key, msg)

  /** HMAC-SHA-256 with an arbitrary-length key (used by EIP-2333 HKDF). */
  def hmacSha256(key: Array[Byte], msg: Array[Byte]): Array[Byte] =
    hmac(hHmac256Statebytes, hHmac256Init, hHmac256Update, hHmac256Final, Sha256Bytes, key, msg)

  private def hmac(statebytes: MethodHandle, init: MethodHandle, update: MethodHandle, fin: MethodHandle,
                   outLen: Int, key: Array[Byte], msg: Array[Byte]): Array[Byte] = withArena { a =>
    val stateLen = statebytes.invokeWithArguments().asInstanceOf[java.lang.Long].longValue
    val state = a.allocate(stateLen, 16L)
    val ki = init.invokeWithArguments(state, seg(a, key), java.lang.Long.valueOf(key.length.toLong)).asInstanceOf[Integer].intValue
    if ki != 0 then throw new RuntimeException("hmac init failed")
    val ku = update.invokeWithArguments(state, seg(a, msg), java.lang.Long.valueOf(msg.length.toLong)).asInstanceOf[Integer].intValue
    if ku != 0 then throw new RuntimeException("hmac update failed")
    val o = out(a, outLen)
    val kf = fin.invokeWithArguments(state, o).asInstanceOf[Integer].intValue
    if kf != 0 then throw new RuntimeException("hmac final failed")
    read(o, outLen)
  }
